narwhal/primary/src/core.rs

```diff
@@ -185,327 +173,260 @@ impl Core {
         self.highest_received_round = last_round_number;
         self.highest_processed_round = last_round_number;
 
-        // Get latest header from all peers and process it.
-        let peers: Vec<NetworkPublicKey> = self
-            .committee
-            .others_primaries(&self.name)
-            .into_iter()
-            .map(|(_, _, network_key)| network_key)
-            .collect();
-
-        let mut header_futures = FuturesUnordered::new();
-
-        for peer in peers.iter() {
-            let network = self.network.network();
-            let request = anemo::Request::new(LatestHeaderRequest {})
-                .with_timeout(LATEST_HEADER_REQUEST_TIMEOUT);
+        self
+    }
 
-            header_futures.push(async move { network.get_latest_header(peer, request).await });
-        }
+    // Requests a vote for a Header from the given peer. Retries indefinitely until either a
+    // vote is received, or a permanent error is returned.
+    #[instrument(level = "debug", skip_all, fields(header_digest = ?header.digest()))]
+    async fn request_vote(
+        network: anemo::Network,
+        committee: Committee,
+        certificate_store: CertificateStore,
+        target: NetworkPublicKey,
+        header: Header,
+    ) -> DagResult<Vote> {
+        let peer_id = anemo::PeerId(target.0.to_bytes());
+        let peer = network.waiting_peer(peer_id);
+        let mut client = PrimaryToPrimaryClient::new(peer);
+
+        let mut missing_parents: Vec<CertificateDigest> = Vec::new();
+        let mut attempt: u32 = 0;
+        let vote = loop {
+            attempt += 1;
+
+            let parents = if missing_parents.is_empty() {
+                Vec::new()
+            } else {
+                let expected_count = missing_parents.len();
```

Comment:
I would propose doing a check here that the other node is only requesting the actual header's parents, and not just random digests (which are not the parents of the header in question).

Reply:
Good suggestion, done